# Домашняя работа №3 по дискретной математике
## Выполнил Анненков Владислав

#### Объявлю функции Python для наглядного счёта:

In [1]:
import math
import pandas as pd


def C(k, n):
    return math.factorial(n) // (math.factorial(k) * math.factorial(n - k))


def p_list(n):
    i = 2
    primfac = []
    while i * i <= n:
        while n % i == 0:
            primfac.append(i)
            n = n // i
        i = i + 1
    if n > 1:
        primfac.append(n)
    return primfac

### Задача №1
Сколько различных натуральных делителей у числа 1224?

Сначала разложим число на простые множители: $1224 = 2^3 \cdot 3^2 \cdot 17$,<br>
тогда все делители такого числа будут иметь вид $2^x \cdot 3^y \cdot 17^z : x \in \{ 0, 1, 2, 3 \}, y \in \{ 0, 1, 2 \}, z \in \{ 0, 1 \}$.

Количество различных комбинаций $x, y, z$ — искомое количество натуральных делителей.<br>
$4 \cdot 3 \cdot 2 = 24$ делителя

__Ответ:__ 24 делителя.

In [2]:
p_list(1224)

[2, 2, 2, 3, 3, 17]

### Задача №2
Сколько есть способов разложить 7 _различных_ монет по 3 _различным_ карманам?

Рассмотрим 9 элементов: 7 монет и 2 "разделения" между тремя карманами.<br>
Тогда остаётся рассмотреть различные комбинации этих 9 элементов:<br>
$C_9^2 = C_9^7 = |\mathcal{P}_2(\underline{9})| = 36$

__Ответ:__ 36 способов.

### Задача №3
Рассмотрим слово $\alpha = cbcad$.<br>
Сколько слов длины от 1 до 5, составленных из букв латинского алфавита (и необязательно осмысленных), стояли бы прежде $\alpha$ в словаре?

В английском алфавите 26 букв.<br>
Тогда $\underline{2 + 2 \cdot (C_{26}^1 + C_{26}^2 + C_{26}^3 + C_{26}^4) = 35804}$ — количество вариантов,<br>
которые начинаются с букв $a$ и $b$.

Рекурсивно рассмотрим варианты, которые начинаются с буквы $c$, но стоят раньше в словаре:<br>
$\underline{1 + 1 \cdot (C_{26}^1 + C_{26}^2 + C_{26}^3) = 2952}$ — количество слов вида $cxyyy \lt \alpha : x \in \{ a \}, y \in \{ a, b, \ldots, y, z \}$.

Продолжаем логическую цепочку по аналогии:<br>
$\underline{2 + 2 \cdot (C_{26}^1 + C_{26}^2) = 54}$ — количество слов вида $cbxyy \lt \alpha : x \in \{ a, b \}, y \in \{ a, b, \ldots, y, z \}$;<br>
$\underline{3}$ — количество слов вида $cbcax \lt \alpha : x \in \{ a, b, c \}$.

Посчитаем общее количество: $\underline{35804 + 2952 + 54 + 3 = 38813}$

__Ответ:__ 38813 слов.

In [3]:
print('1:', 2 + 2 * (C(1, 26) + C(2, 26) + C(3, 26) + C(4, 26)))
print('2:', 1 + 1 * (C(1, 26) + C(2, 26) + C(3, 26)))
print('3:', 2 + 2 * (C(1, 26)))
print('4:', 0)
print('5:', 3)

1: 35804
2: 2952
3: 54
4: 0
5: 3


In [4]:
print('Ответ:', 35804 + 2952 + 54 + 3)

Ответ: 38813


### Задача №5
Каждую секунду робот делает один шаг направо или один шаг вверх. Сколькими различными путями он может попасть в точку $(m, n)$ из точки $(0, 0)$ при $m, n \in N$? (Координаты понимаются естественным образом.)

Суммарное количество шагов робота $\alpha = m + n$.

Пусть есть последовательность $\{ 1, 2, \ldots, \alpha \}$ — наши ходы с неопределёнными движениями,<br>
где каждый элемент может быть либо движением вверх, либо вправо.<br>
Тогда мы можем взять различные $m$ элементов и принять их за движения вправо,<br>
тогда оставшиеся $n$ будут движениями вверх.

Мы можем располагать движения робота в любом порядке, главное — их сумма должна быть равна $\alpha$,<br>
поэтому количество различных путей — это количество различных подмножеств из $m$ (или $n$) элементов множества $\{ 1, 2, \ldots, \alpha \}$.

Количество путей:<br>
$$C_{m + n}^m = |\mathcal{P}_m(\underline{m + n})| = \frac{(m + n)!}{m! \cdot ((m + n) - m)!} = \frac{(m + n)!}{m! \cdot n!}$$

Это будет __ответ__.

### Задача №6
Рассмотрим $n \gt 1$ идущих подряд целых чисел: $a, a+1, \ldots, a + (n - 1)$. Докажите, что их произведение всегда кратно $n!$.

По определению биноминального коэффициента $C^n_{n + k}$ — целое число.<br>
$\frac{(n + k)!}{n!(n + k - n)!} = \frac{(n + k)!}{n!k!} = \frac{(n + 1)(n + 2) \ldots (n + k)}{k!} \Rightarrow$ все $k$ подряд идущих чисел делятся на $k!$.

Примем $a = n + 1$, а $k = n$ и получим выражение $\frac{a(a + 1)(a + 2) \ldots (a + (n - 1))}{n!}$, которое всегда делится нацело.

Ч.Т.Д.

### Задача №7
Сколько существует шестизначных чисел, в записи которых поровну четных и нечетных цифр?

Для начала посчитаем количество чисел с различными расстановками чётно-нечётных цифр.<br>
Очевидно, что количество чётных = количество нечётных = 3, тогда количество видов таких чисел: $C_6^3 = 20$.

В каждом таком числе есть набор цифр. Существует 5 чётных цифр и 5 нечётных.<br>
Тогда комбинаций 5 цифр размера 3 $= 5^3 = 125$. Это количество возможных комбинаций для чётных и нечётных цифр.

Суммарное количество комбинаций $= 125 + 125 = 250$.<br>
То есть, каждое число конкретного вида имеет 250 своих вариаций, где:<br>
- Вид числа: различные комбинации $x, y$ размера 6, т.ч. кол-во $x$ = кол-во $y$ = $3$, и $x \in \{ 0, 2, 4, 6, 8 \}, y \in \{ 1, 3, 5, 7, 9 \}$ (при этом все $x$ и $y$ могут быть различны),<br>напр.: $xxxyyy, xyxyxy, xxyyyx, \ldots$;
- Вариация числа: набор конкретных значений цифр для числа определённого вида.

__Ответ:__ $20 \cdot 250 = 5000$ вариантов.

### Задача №8
Сколькими способами можно разделить 6 яблок, 3 груши и 2 сливы между Иваном, Петром, Анной и Дарьей, если фрукты одного вида не различаются?

Разделим Ивана, Петра, Анну и Дарью мысленными разделителями. Пусть эти разделители — часть наших множеств с яблоками, грушами и т.д.<br>
Тогда посчитаем возможные способы для множеств каждого вида по-отдельности.

- яблоки: $C^4_{6 + 4} = 210$;
- груши: $C^4_{3 + 4} = 35$;
- сливы: $C^4_{2 + 4} = 15$.

Для получения ответа достаточно перемножить эти значения.

__Ответ:__ $210 \cdot 35 \cdot 15 = 110250$.

### Задача №9
Найдите коэффициент при $x^{57}$ в многочлене $(x^2 + x^7 + x^9)^{20}$.

$(x^2 + x^7 + x^9)^{20} = x^{40}(1 + x^5 + x^7)^{20}$

Требуется получить $x^{17}$.<br>
Из комбинаций пятёрки и семёрки это возможно сделать единственным способом: $5 + 5 + 7 = 17$.

Следовательно, в биноминальном разложении скобки, где $a = 1 + x^5, b = x^7$, требуется слагаемое $C^1_{20}a^{19}b = C^1_{20}(1 + x^5)^{19}x^7$.

Очевидно, теперь требуется найти коэффициент при $x^{10}$ в разложении $(1 + x^5)^{19} = C^2_{19} = 171$.

Тогда коэффициент при $x^{57} = C^1_{20} \cdot C^2_{19} = 171 \cdot 20 = 3420$.

__Ответ:__ 3420.

### Задача №10
Сколькими способами можно разложить семь различных монет по трем различным карманам так, чтобы каждый карман был не пуст?

Рассмотрим пространства между монетами. В каждое такое пространство можно вставить мысленную "перегородку",<br>
которая будет отделять монеты каждого кармана. Таких пространств будет 6, и в эти пространства мы можем<br>
каким-то образом вставить 2 перегородки. Тогда в каждый карман обязательно попадёт хотя бы одна монета.

__Ответ:__ $C_6^2 = 15$ способов.

### Задача №11
Скольким способами можно переставить 10 стоящих в ряд книг, чтобы ровно 4 из них остались на месте?

Общий алгоритм следующий: убрать какие-то 4 книги, а количество остальных посчитать факториалом.<br>
Тогда для конкретного набора 4-х книг число перестановок будет: $(10-4)! = 6! = 1 \cdot 2 \cdot 3 \cdot 4 \cdot 5 \cdot 6 = 720$.

Тогда посчитаем количество различных наборов по 4 книги: $C_{10}^4 = 210$.

Всего таких перестановок: $720 \cdot 210 = 151200$.

__Ответ:__ 151200 способов.

### Задача №12
Сколько чисел из 2020 _не_ взаимно просты с 2020?

$2020 = 2^2 \cdot 5 \cdot 101 \Rightarrow \phi(2020) = \phi(2^2) \cdot \phi(5) \cdot \phi(101) = 800$.

800 чисел взаимно просты, тогда $2020 - 800 = 1220$ не взаимно просты.

__Ответ:__ 1220.

### Задача №13
Сколько есть семизначных чисел $a_7a_6 \ldots a_1$, где $a_i \leq a_{i + 1}$ при всех $i$?

Попробуем рекурсивно построить таблицу следующим образом.

__Шаг 1.__<br>
Построим мысленно первый (самый правый) столбец со строчками-цифрами 0-9. Он будет содержать одни единицы, потому что мы можем взять любую цифру, удовлетворяющую условию задачи, если такое число однозначное.

__Шаг 2.__<br>
Теперь построим второй столбец слева. Заполним его по принципу: каждая ячейка это сумма всех возможных значений справа.<br>
Например, с самым верхним элементом могут быть любые 10 элементов справа, с элементом на второй строке — любые 9 и т.д.<br>
$\begin{pmatrix}
    \Sigma^{10}_{i = 1} & 1\\
    \Sigma^{9}_{i = 1} & 1\\
    \Sigma^{8}_{i = 1} & 1\\
    \Sigma^{7}_{i = 1} & 1\\
    \Sigma^{6}_{i = 1} & 1\\
    \Sigma^{5}_{i = 1} & 1\\
    \Sigma^{4}_{i = 1} & 1\\
    \Sigma^{3}_{i = 1} & 1\\
    \Sigma^{2}_{i = 1} & 1\\
    \Sigma^{1}_{i = 1} & 1\\
\end{pmatrix} \Rightarrow
\begin{pmatrix}
    10 & 1\\
    9 & 1\\
    8 & 1\\
    7 & 1\\
    6 & 1\\
    5 & 1\\
    4 & 1\\
    3 & 1\\
    2 & 1\\
    1 & 1\\
\end{pmatrix}$

__Шаг 3.__
По такому же принципу строим третий столбец, только теперь суммируем элементы из второго столбца.<br>
$\begin{pmatrix}
    55 & 10 & 1\\
    45 & 9 & 1\\
    36 & 8 & 1\\
    28 & 7 & 1\\
    21 & 6 & 1\\
    15 & 5 & 1\\
    10 & 4 & 1\\
    6 & 3 & 1\\
    3 & 2 & 1\\
    1 & 1 & 1\\
\end{pmatrix}$

__Шаг 4-n.__
Достраиваем остальные столбцы по такому же принципу. Для экономии времени и точности я написал код на Python.

In [5]:
result = [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

for i in range(6):
    result.append(list())
    for l in range(10):
        s = 0
        for k in range(l, len(result[i])):
            s += result[i][k]

        result[i + 1].append(s)

pd.DataFrame(result)

0     1     2    3    4    5   6   7  8  9
0     1     1     1    1    1    1   1   1  1  1
1    10     9     8    7    6    5   4   3  2  1
2    55    45    36   28   21   15  10   6  3  1
3   220   165   120   84   56   35  20  10  4  1
4   715   495   330  210  126   70  35  15  5  1
5  2002  1287   792  462  252  126  56  21  6  1
6  5005  3003  1716  924  462  210  84  28  7  1

Получившаяся таблица выше. Единственное — по техническим причинам она выведена не столбцами, а строками, будто счёт вёлся сверху вниз, добавляя каждый раз новую строчку. Но на ответ это не повлияло.

Количество таких чисел — левое нижнее значение (в данной таблице).

__Ответ:__ 5005.